firestore-send-email/functions/src/index.ts

```diff
@@ -278,76 +272,146 @@ async function deliver(
+          },
+        } as QueuePayload);
+        // We've updated the payload, so we need to attempt delivery, but we
+        // don't want to do it in this transaction. Since the transaction will
+        // update the record again the cloud function will be triggered again
+        // and delivery will be attempted at that point.
+        return false;
+      }
+
+      // The email has already been delivered, so we don't need to do anything.
+      if (payload.delivery.state === "SUCCESS") {
+        return false;
+      }
+
+      // The email has previously failed to be delivered, so we can't do anything.
+      if (payload.delivery.state === "ERROR") {
+        return false;
+      }
 
-  switch (payload.delivery.state) {
-    case "SUCCESS":
-    case "ERROR":
-      return null;
-    case "PROCESSING":
-      if (payload.delivery.leaseExpireTime.toMillis() < Date.now()) {
-        // Wrapping in transaction to allow for automatic retries (#48)
-        return admin.firestore().runTransaction((transaction) => {
-          transaction.update(change.after.ref, {
+      if (payload.delivery.state === "PROCESSING") {
+        if (payload.delivery.leaseExpireTime.toMillis() < Date.now()) {
+          // The lease has expired, so we should not attempt to deliver the email again,
+          // but we set the state to ERROR so clients can see that the email failed.
+          transaction.update(ref, {
             "delivery.state": "ERROR",
             // Keeping error to avoid any breaking changes in the next minor update.
             // Error to be removed for the next major release.
             error: "Message processing lease expired.",
             "delivery.error": "Message processing lease expired.",
           });
-          return Promise.resolve();
-        });
+        }
+        // Already being processed, so we don't need to do anything.
+        return false;
       }
-      return null;
-    case "PENDING":
-    case "RETRY":
-      // Wrapping in transaction to allow for automatic retries (#48)
-      await admin.firestore().runTransaction((transaction) => {
-        transaction.update(change.after.ref, {
+
+      if (
+        payload.delivery.state === "PENDING" ||
+        payload.delivery.state === "RETRY"
+      ) {
+        // We can attempt to deliver the email in these states, so we set the state to PROCESSING
+        // and set a lease time to prevent delivery from being attempted forever.
+        transaction.update(ref, {
           "delivery.state": "PROCESSING",
           "delivery.leaseExpireTime": admin.firestore.Timestamp.fromMillis(
             Date.now() + 60000
           ),
         });
-        return Promise.resolve();
-      });
-      return deliver(payload, change.after.ref);
+        return true;
+      }
+
+      // We don't know what the state is, so we can't do anything. This should never happen.
+      return false;
+    });
+
+  if (shouldAttemptDelivery) {
+    await deliver(ref);
```

Comment:
If deliver fails, should we try/catch that and move delivery.state to RETRY?

Reply:
The deliver method has its own error handling and sets either SUCCESS or ERROR states. Setting it to RETRY on failure would be a extension behaviour change, since currently to RETRY a user must explicitly set the state to RETRY themselves external to the extension.